### DISCUSSION TITLE
Buckets in prometheus.HistogramsOpts in golang shows as undefined while accessing.

### DISCUSSION THREAD

**Discussion by arjunbakshi245** [prometheus/prometheus#12295] [Q&A] [answered]
I have defined a histogram metric in prometheus with my custom buckets. But while running a gatherer on all my metrics the Buckets field for the histogram says 
`mf.Buckets undefined (type *io_prometheus_client.MetricFamily has no field or method Buckets)`
Need help to access the length of the Buckets field in the my custom gatherer

**Comment by beorn7** [comment]
`Bucket` is a (repeated) field of the `Histogram` message, which itself is a field in the `Metric` message, which itself is a field in the `MetricFamily` message. Or in other words: You have to go two levels deeper.

**Reply by arjunbakshi245** [reply]
@beorn7 By two levels do you mean that the buckets is defined for a metric and not a metric family? so I must do something similar to the below snippet? Because this throws the same error of 
### m.Buckets undefined (type *io_prometheus_client.Metric has no field or method Buckets)
```
for _, mf := range mfs{
        if mf.GetType() == dto.MetricType_HISTOGRAM{
                for _, m := range mf.GetMetric() {
	             fmt.Println(m.Buckets)
                } 
       }
}
```

**Reply by arjunbakshi245** [reply]
I want to print out the buckets array { 0, 0.25, 0.5 .....} like so for each histogram.

**Reply by beorn7** [reply, marked as answer]
That would be one level deeper (MetricFamily → Metric), but you have to go two levels deeper (Metric → Histogram).

**Reply by arjunbakshi245** [reply]
How exactly can I do that?

**Reply by arjunbakshi245** [reply]
@beorn7  This is how the family is defined.
```
recover_bytes = prometheus.NewHistogram(prometheus.HistogramOpts{
		Name:    "abc",
		Help:    "xyz",
		Buckets: []float64{0.1, 0.5, 1, 2, 4, 8},
	})
```
I am able to print **mf.Name** and **mf.Help**. Just not the Buckets.

**Reply by beorn7** [reply]
You have to look at the actual Go code generated from the proto spec. Maybe your IDE will support you with that. You already wrote `range mf.GetMetric()` above. It will iterate through the metrics. On each metric, you can call `m.GetHistogram()`, which should then allow you to access buckets.

**Comment by beorn7** [comment]
BTW: This is more a question about prometheus/client_golang, not prometheus/prometheus.

**Reply by arjunbakshi245** [reply]
@beorn7 Thanks a lot for the answer. Will post on prometheus/client_golang next time. Have a great day
